### ISSUE TITLE
Why does QuickCheck.GenT.infiniteListOf diverge in this example?

### ISSUE THREAD

**Issue by runeksvendsen** [nikita-volkov/QuickCheck-GenT#8]
Hi there,

Thank you for creating this useful library.

While using the library, I encountered something I don't quite understand, so I thought I'd ask here for clarification.

Consider the following code:

```haskell
import qualified QuickCheck.GenT as GenT
import qualified Test.QuickCheck as Gen
import Control.Monad (join)


infiniteListBad, infiniteListGood :: Gen.Gen a -> GenT.GenT IO [a]

infiniteListBad = GenT.infiniteListOf . GenT.liftGen
infiniteListGood = GenT.liftGen . Gen.infiniteListOf

toIO :: GenT.GenT IO a -> IO a
toIO = join . Gen.generate . GenT.runGenT

diverges :: IO ()
diverges = do
    list <- toIO (infiniteListBad $ return 42)
    print $ head list

prints42 :: IO ()
prints42 = do
    list <- toIO (infiniteListGood $ return 42)
    print $ head list
```

Here, `prints42` prints "42" when evaluated, while `diverges` diverges. Why does this happen?

**Comment by nikita-volkov** [comment]
Those functions were not implemented by me (see https://github.com/nikita-volkov/QuickCheck-GenT/commit/cde49fbd679ef89d3272b757c2cee8e78b1d820c). Let's ask @brandon-leapyear.

**Comment by brandon-leapyear** [comment]
:sparkles: _**This is an old work account. Please reference @brandonchinn178 for all future communication**_ :sparkles:


---

Still investigating. So far, I found that `infiniteListOf . liftGen` works for `GenT Identity [a]`, so it's something about `IO`

**Comment by brandon-leapyear** [comment]
:sparkles: _**This is an old work account. Please reference @brandonchinn178 for all future communication**_ :sparkles:


---

Ok so `infiniteListOf` is implemented as `sequence . repeat`, and it seems like that isn't good for infinite lists in IO (but it works for infinite lists in other monads like `Identity` or `Gen`). The following still hangs:
```hs
list <- sequence $ repeat $ pure 42
print $ head list
```

**Comment by runeksvendsen** [comment]
Yes, I also see a difference between monads. However, `GenT []` in the below example also diverges for me, so it appears it's not just `IO`:

```haskell
import qualified QuickCheck.GenT as GenT
import qualified Test.QuickCheck as Gen


infiniteListBad, infiniteListGood :: Gen.Gen a -> GenT.GenT [] [a]

infiniteListBad = GenT.infiniteListOf . GenT.liftGen
infiniteListGood = GenT.liftGen . Gen.infiniteListOf

diverges :: IO ()
diverges = do
    list <- Gen.generate $ GenT.runGenT (infiniteListBad $ return 42)
    print $ head $ head list

prints42 :: IO ()
prints42 = do
    list <- Gen.generate $ GenT.runGenT (infiniteListGood $ return 42)
    print $ head $ head list
```

**Comment by runeksvendsen** [comment]
It looks like the deciding factor is whether the given monad is strict or lazy. Using `Control.Monad.State.Lazy.State s` there is no divergence, whereas `GenT.infiniteListOf` diverges when using `Control.Monad.State.Strict.State s`.

**Comment by brandon-leapyear** [comment]
:sparkles: _**This is an old work account. Please reference @brandonchinn178 for all future communication**_ :sparkles:


---

Yes, exactly: https://stackoverflow.com/a/14496898/8565175.

Not sure what the best route forward is. Maybe change
```hs
infiniteListOf :: MonadGen m => m a -> m [a]
```
to
```hs
infiniteListOf :: MonadGen m => Gen a -> m [a]
```
? Does that still work for your use-case?

**Comment by runeksvendsen** [comment]
I'm not sure either what the best course of action is. But your suggestion certainly works for my use case. As I see it, we can't require the monad of the generator to be lazy if its type is just `MonadGen m => m a`, so there's no other way to solve it.

Side note: an article of @nikita-volkov seems very relevant: https://nikita-volkov.github.io/if-haskell-were-strict/. Since this is exactly what we're looking for here: a way to specify, through the type system, that a given monad must be lazy.
